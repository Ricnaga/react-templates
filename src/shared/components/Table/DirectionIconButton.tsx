import { useMemo } from 'react';

import { TriangleDownIcon, TriangleUpIcon, UpDownIcon } from '@chakra-ui/icons';
import { IconButton } from '@chakra-ui/react';

interface DirectionIconButtonProps {
  onChangeDirection: () => void;
  isSelectedField: boolean;
  isAscending: boolean;
}

export function DirectionIconButton({
  onChangeDirection,
  isSelectedField,
  isAscending,
}: DirectionIconButtonProps) {
  const icon = useMemo(() => {
    if (isSelectedField) return <UpDownIcon />;
    return isAscending ? <TriangleUpIcon /> : <TriangleDownIcon />;
  }, [isAscending, isSelectedField]);

  return (
    <IconButton
      aria-label="Order Icon"
      icon={icon}
      variant="link"
      onClick={onChangeDirection}
    />
  );
}
