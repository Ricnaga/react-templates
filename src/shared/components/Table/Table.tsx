import { ReactNode, useState } from 'react';

import {
  Table as ChakraTable,
  TableCaption,
  TableContainer,
  Tbody,
  Th,
  Thead,
  Tr,
} from '@chakra-ui/react';

import { DirectionIconButton } from './DirectionIconButton';

type Data = Record<symbol, symbol>;

export type TableHeaderKeys<T extends Data, Extra extends string = ''> = Array<{
  header: keyof T | Extra;
  title: string | ReactNode;
  colSpan?: number;
}>;

interface TableProps<T extends Data, Extra extends string = ''> {
  header: TableHeaderKeys<T, Extra>;
  data: Array<T>;
  children: (data: T) => ReactNode;
}

enum OrderBy {
  ASC = 'ASC',
  DESC = 'DESC',
}

export function Table<T extends Data, Extra extends string = ''>({
  data,
  header,
  children,
}: TableProps<T, Extra>) {
  const [sortField, setSortField] = useState<keyof T>();
  const [order, setOrder] = useState<keyof typeof OrderBy>('ASC');

  const onOrderBy = (fieldName: keyof T) => {
    const sortOrder =
      fieldName === sortField && order === 'ASC' ? OrderBy.DESC : OrderBy.ASC;
    setSortField(fieldName);
    setOrder(sortOrder);

    data.sort((a, b) => {
      const compareFields = a[fieldName] > b[fieldName];

      if (sortOrder === OrderBy.ASC) return compareFields ? 1 : -1;
      if (sortOrder === OrderBy.DESC) return compareFields ? -1 : 1;
      return 0;
    });
  };

  return (
    <TableContainer>
      <ChakraTable variant="simple">
        <Thead>
          <Tr>
            {header.map((value) => (
              <Th key={value.header.toString()} colSpan={value.colSpan ?? 1}>
                {value.title}{' '}
                <DirectionIconButton
                  isSelectedField={sortField !== value.header}
                  isAscending={order === 'ASC'}
                  onChangeDirection={() => onOrderBy(value.header as keyof T)}
                />
              </Th>
            ))}
          </Tr>
        </Thead>
        <Tbody>
          {data.map((value) => (
            <Tr key={Math.random().toString()}>{children(value)}</Tr>
          ))}
        </Tbody>
        <TableCaption>Exemplo de tabela ordenada</TableCaption>
      </ChakraTable>
    </TableContainer>
  );
}
