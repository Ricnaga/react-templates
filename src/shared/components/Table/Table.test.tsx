import { Td } from '@chakra-ui/react';

import { render, userEvent, within } from '@application/test/testing-library';

import { Table, TableHeaderKeys } from './Table';

const header: TableHeaderKeys<{ id: string }> = [
  { header: 'id', title: 'Mock id' },
];

const data: Array<{ id: string }> = [
  { id: 'id1' },
  { id: 'id1' },
  { id: 'id2' },
];

describe('Component: Table', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Table header={header} data={data}>
        {({ id }) => <Td>{id}</Td>}
      </Table>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should order column', async () => {
    const { getByRole } = render(
      <Table header={header} data={data}>
        {({ id }) => <Td>{id}</Td>}
      </Table>,
    );
    const columnheader = getByRole('columnheader', {
      name: /mock id/i,
    });
    const columnButton = within(columnheader).getByRole('button', {
      name: /order icon/i,
    });

    await userEvent.click(columnButton);
    await userEvent.click(columnButton);
    await userEvent.click(columnButton);
    await userEvent.click(columnButton);
    expect(data[0].id).toBe('id2');
  });
});
