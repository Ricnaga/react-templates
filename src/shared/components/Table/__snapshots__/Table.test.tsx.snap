// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Table should render correctly 1`] = `
<div>
  <div
    class="chakra-table__container css-zipzvv"
  >
    <table
      class="chakra-table css-5605sr"
    >
      <thead
        class="css-0"
      >
        <tr
          class="css-0"
        >
          <th
            class="css-1zts0j"
          >
            <button
              aria-label="Order Icon"
              class="chakra-button css-1j1nr9i"
              type="button"
            >
              Mock id
              <span
                class="chakra-button__icon css-1hzyiq5"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-icon css-onkibi"
                  focusable="false"
                  viewBox="-1 -1 9 11"
                >
                  <path
                    d="M 3.5 0L 3.98809 -0.569442L 3.5 -0.987808L 3.01191 -0.569442L 3.5 0ZM 3.5 9L 3.01191 9.56944L 3.5 9.98781L 3.98809 9.56944L 3.5 9ZM 0.488094 3.56944L 3.98809 0.569442L 3.01191 -0.569442L -0.488094 2.43056L 0.488094 3.56944ZM 3.01191 0.569442L 6.51191 3.56944L 7.48809 2.43056L 3.98809 -0.569442L 3.01191 0.569442ZM -0.488094 6.56944L 3.01191 9.56944L 3.98809 8.43056L 0.488094 5.43056L -0.488094 6.56944ZM 3.98809 9.56944L 7.48809 6.56944L 6.51191 5.43056L 3.01191 8.43056L 3.98809 9.56944Z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-0"
        >
          <td
            class="css-xumdn4"
          >
            id1
          </td>
        </tr>
        <tr
          class="css-0"
        >
          <td
            class="css-xumdn4"
          >
            id1
          </td>
        </tr>
        <tr
          class="css-0"
        >
          <td
            class="css-xumdn4"
          >
            id2
          </td>
        </tr>
      </tbody>
      <caption
        class="css-7478fk"
      >
        Exemplo de tabela ordenada
      </caption>
    </table>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
